Context policy variable evaluation

cianfoley-nearform opened the pull request:
@dberesford @mihaidma 

I rebased this branch with master, the context stuff is in here and I added some testing around it, as well as tests around not allowing reserved keywords as variable names in policy instances.

Volume db and autocannon volume tests can be run now from hapi-server-plugin module too.

mihaidma commented on packages/udaru-core/lib/ops/authorizeOps.js:
```diff
@@ -19,6 +19,20 @@ function badImplementationWrap (next) {
   }
 }
 
+function getContext (params) {
```
small comment on naming: this looks more like a `buildContext`

cianfoley-nearform replied:
will change :-)

dberesford commented on packages/udaru-core/lib/ops/authorizeOps.js:
```diff
@@ -19,6 +19,20 @@ function badImplementationWrap (next) {
+    },
+    request: {
+      currentTime: new Date().toISOString(),
+      sourceIp: params.sourceIpAddress,
```
in terms of validation, these can be null, so I would only add them to the request object if they're not null (I wouldn't let params through with null values to be evaluated in contexts)

cianfoley-nearform replied:
Yes i was thinking this, i'm not sure what it would evaluate to, but was thinking policies could have conditions where it does one thing if accessed with no ip address e.g. through core, vs via a serving API... maybe a separate context variable would be good for this purpose

e.g. request.package = udaru-core

dberesford commented on packages/udaru-core/package.json:
```diff
@@ -44,6 +44,7 @@
     "depcheck": "npx depcheck",
     "pg:init": "node ./database/init.js && npm run pg:migrate",
     "pg:init-test-db": "npm run pg:init && npm run pg:load-test-data",
+    "pg:init-volume-db": "npm run pg:init && node ./database/loadVolumeData.js",
```
we seem to have this in `udaru-core` and `udaru-hapi-server`, is there a good reason it's in core?

cianfoley-nearform replied:
possibly not, it was just in there so I left it... it could be useful for us to have some tests with the volume data set but there are none at the moment so it could probably be moved out altogether... currently it's only used for the bench tests but I've found it handy for testing manually

dberesford replied:
Let's remove from core and keep it all in server, thanks

mihaidma commented on packages/udaru-core/test/integration/authorizeOps.test.js:
```diff
@@ -181,6 +181,357 @@ lab.experiment('AuthorizeOps', () => {
+  lab.test('authorize isUserAuthorized - test with udaru:organizationId variable for condition', (done) => {
+    let Condition = { StringEquals: { 'udaru:organizationId': organizationId } }
+
+    let testPolicy = {
```
linter should complain about the two let's from above

cianfoley-nearform replied:
never complained when I ran lint etc.  I changed in latest commit in any case

mihaidma commented:
@cianfoley-nearform not exactly the main target of this pr but should we also have an e2e test showing the usage of conditionals for org for example?

cianfoley-nearform commented:
@mihaidma @dberesford 

- added ip context to 2 other functions in authorize (added extra context variable called request.source, which is set to 'server' if IP address is present, otherwise it's set to 'api')
- added some documentation for the variables in the core documentation
- added some documentation for the volume and bench in the udaru-hapi-server documentation
- extra tests in core and server around IP address and other things

mihaidma commented:
@cianfoley-nearform LGTM

dberesford commented on docs/overview.md:
```diff
@@ -141,11 +144,42 @@ The main elements of a Statement are:
 -   The Action to be performed. Example: `'authorization:teams:create'` or `'authorization:organizations:*'`
 -   The Resource on which the action is performed. A Resource name is effectively a URI for your resources. Example: `'FOO:orga:CLOUDCUCKOO:scenario:*:entity:north-america-id'`
 -   The Effect - has the value 'Allow' or 'Deny'.
+-   The Conditions, which contain extra logic to determine whether access is granted to the resources
```
Need to stress here that the condition field is optional, e.g. 'optional extra logic'...

dberesford commented on docs/overview.md:
```diff
@@ -141,11 +144,42 @@ The main elements of a Statement are:
 
 For a detailed description of Policies, see the [AWS Policy Elements Reference](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements.html).
 
+## Policy Context Variables & Conditions
```
this is great
